proxy/router/src/lib.rs

```diff
@@ -116,42 +134,55 @@ where T: Recognize,
     type Error = Error<T::Error, T::RouteError>;
     type Future = ResponseFuture<T>;
 
+    /// Always ready to serve.
+    ///
+    /// Graceful backpressure is **not** supported at this level, since each request may
+    /// be routed to different resources. Instead, requests should be issued and each
+    /// route should support a queue of requests.
+    ///
+    /// TODO Attempt to free capacity in the router.
     fn poll_ready(&mut self) -> Poll<(), Self::Error> {
         Ok(().into())
     }
 
+    /// Routes the request through an underlying service.
+    ///
+    /// The response fails when the request cannot be routed.
     fn call(&mut self, request: Self::Request) -> Self::Future {
         let inner = &mut *self.inner.lock().expect("lock router cache");
 
-        let key = match inner.recognize.recognize(&request) {
-            Some(k) => k,
-            None => {
-                return ResponseFuture { state: State::NotRecognized };
-            }
-        };
+        match inner.recognize.recognize(&request) {
+            None => ResponseFuture::not_recognized(),
 
-        // Is the bound service for that key reusable? If `recognize`
-        // returned `SingleUse`, that indicates that the service may
-        // not be used to serve multiple requests.
-        if let Reuse::Reusable(ref k) = key {
-            if let Some(service) = inner.routes.get_mut(k) {
-                let response = service.call(request);
-                return ResponseFuture { state: State::Inner(response) };
+            Some(Reuse::SingleUse(key)) => {
+                // TODO Keep SingleUse services in the cache as well, so that their
+                // capacity is considered. To do this, we should move the Reuse logic into
+                // the returned service (and not the key).
+                let mut service = try_bind_route!(inner.recognize.bind_service(&key));
+                ResponseFuture::new(service.call(request))
             }
-        };
 
-        let mut service = match inner.recognize.bind_service(key.as_ref()) {
-            Ok(s) => s,
-            Err(e) => {
-                return ResponseFuture { state: State::RouteError(e) };
-            }
-        };
+            Some(Reuse::Reusable(key)) => {
+                // First, try to load a cached rute for `key`.
+                if let Some(service) = inner.routes.get_mut(&key) {
+                    return ResponseFuture::new(service.call(request));
+                }
+
+                // Since there wasn't a cached route, ensure that there is capacity for a
+                // new one.
+                if inner.routes.len() == inner.capacity {
+                    // TODO If the cache is full, evict the oldest inactive route. If all
+                    // routes are active, fail the request.
```

Comment:
Out of curiosity: do we believe that LRU is going to be the best eviction strategy here? 

If we have a concept of "active" and "inactive", I'm wondering if we want to just evict the _first_ inactive route we find. That way we don't have to do any bookkeeping of when services were added to the cache (and eliminates the syscalls implied by timestamping them). 

Might be worth testing this?

Reply:
tbd